File: packages/core/src/quiz-engine.ts
import type { Question } from '@aws-prep/content';
import type { Session, AnswerEntry, Strategy } from './types.js';

export function selectQuestions(
  all: Question[],
  strategy: Strategy,
  options: { examId?: number; n?: number; weakTopics?: string[] } = {}
): Question[] {
  const { examId, n, weakTopics = [] } = options;
  let pool: Question[];

  switch (strategy) {
    case 'sequential':
      pool = examId !== undefined ? all.filter(q => q.examId === examId) : all;
      break;
    case 'random':
      pool = [...all].sort(() => Math.random() - 0.5);
      break;
    case 'weak-topic': {
      const filtered = weakTopics.length > 0
        ? all.filter(q => q.topics.some(t => weakTopics.includes(t)))
        : [];
      pool = filtered.length > 0
        ? filtered.sort(() => Math.random() - 0.5)
        : [...all].sort(() => Math.random() - 0.5);
      break;
    }
  }

  return n !== undefined ? pool.slice(0, n) : pool;
}

export function gradeAnswer(question: Question, picked: string[]): boolean {
  if (picked.length === 0) return false;
  const correct = new Set(question.correctLetters);
  const chosen = new Set(picked);
  if (correct.size !== chosen.size) return false;
  for (const c of correct) if (!chosen.has(c)) return false;
  return true;
}

export function createSession(
  questions: Question[],
  strategy: Strategy,
  examId?: number
): Session {
  return {
    id: globalThis.crypto.randomUUID(),
    strategy,
    examId,
    questions,
    currentIdx: 0,
    answers: [],
    startedAt: Date.now(),
    timePerQ: [],
  };
}

export function submitAnswer(session: Session, picked: string[], timeMs: number): Session {
  const q = session.questions[session.currentIdx];
  if (!q) return session;
  const entry: AnswerEntry = {
    examId: q.examId,
    questionNumber: q.number,
    picked,
    correct: gradeAnswer(q, picked),
    timeMs,
  };
  return {
    ...session,
    answers: [...session.answers, entry],
    timePerQ: [...session.timePerQ, timeMs],
  };
}

export function nextQuestion(session: Session): Session {
  return { ...session, currentIdx: session.currentIdx + 1 };
}

export function isSessionComplete(session: Session): boolean {
  return session.currentIdx >= session.questions.length;
}

export function getSessionScore(session: Session): { correct: number; total: number; pct: number } {
  const correct = session.answers.filter(a => a.correct).length;
  const total = session.answers.length;
  return { correct, total, pct: total > 0 ? Math.round((correct / total) * 100) : 0 };
}

File: myXcodeApp/myXcodeApp/Engine/QuizEngine.swift
import Foundation

enum QuizStrategy: String, Codable {
    case sequential, random, weakTopic = "weak-topic"
}

struct AnswerEntry: Codable {
    let questionId: String
    let picked: [String]
    let correct: Bool
    let timeMs: Int
}

struct QuizSession: Identifiable {
    let id: String
    let strategy: QuizStrategy
    let examId: Int?
    let questions: [ContentQuestion]
    var currentIdx: Int
    var answers: [AnswerEntry]
    let startedAt: Date
    var timePerQ: [Int]

    var currentQuestion: ContentQuestion? {
        guard currentIdx < questions.count else { return nil }
        return questions[currentIdx]
    }

    var isComplete: Bool { currentIdx >= questions.count }

    var score: (correct: Int, total: Int, pct: Double) {
        let correct = answers.filter { $0.correct }.count
        let total = answers.count
        let pct = total > 0 ? Double(correct) / Double(total) : 0
        return (correct, total, pct)
    }
}

func selectQuestions(
    all: [ContentQuestion],
    strategy: QuizStrategy,
    examId: Int? = nil,
    weakTopicIds: Set<String> = [],
    limit: Int? = nil
) -> [ContentQuestion] {
    var pool = examId != nil ? all.filter { $0.examId == examId } : all

    switch strategy {
    case .random:
        pool = pool.shuffled()
    case .weakTopic:
        if !weakTopicIds.isEmpty {
            pool = pool.filter { q in q.topics.contains(where: { weakTopicIds.contains($0) }) }
        }
        pool = pool.shuffled()
    case .sequential:
        break
    }

    if let limit = limit {
        pool = Array(pool.prefix(limit))
    }
    return pool
}

func gradeAnswer(question: ContentQuestion, picked: [String]) -> Bool {
    Set(picked) == Set(question.correctLetters)
}

func createSession(questions: [ContentQuestion], strategy: QuizStrategy, examId: Int? = nil) -> QuizSession {
    QuizSession(
        id: UUID().uuidString,
        strategy: strategy,
        examId: examId,
        questions: questions,
        currentIdx: 0,
        answers: [],
        startedAt: Date(),
        timePerQ: []
    )
}

func submitAnswer(session: inout QuizSession, picked: [String], timeMs: Int) {
    guard let q = session.currentQuestion else { return }
    let correct = gradeAnswer(question: q, picked: picked)
    let entry = AnswerEntry(questionId: q.id, picked: picked, correct: correct, timeMs: timeMs)
    session.answers.append(entry)
    session.timePerQ.append(timeMs)
}

func advanceSession(session: inout QuizSession) {
    session.currentIdx += 1
}
